### PR TITLE
New rule: getter-return

### LINKED ISSUE #8449: New Rule: Enforce return in getter
**Please describe what the rule should do:**

Enforces that a return statement is present in property getters

**What category of rule is this? (place an "X" next to just one item)**

[ ] Enforces code style
[X] Warns about a potential error
[ ] Suggests an alternate way of doing something
[ ] Other (please specify:)

**Provide 2-3 code examples that this rule will warn about:**

```js

/*eslint getter-return: "error"*/

// Incorrect
class MyClass {
  get a() {
     this.retreiveVal('a');
  }
}

function MyConstructorFunction() {
  Object.defineProperty(this, 'a', {
    get: function () {
      this.retreiveVal('a');
    }
  });
}


// Correct
class MyClass {
  get a() {
     return this.retreiveVal('a');
  }
}

function MyConstructorFunction() {
  Object.defineProperty(this, 'a', {
    get: function () {
       return this.retreiveVal('a');
    }
  });
}
```

**Why should this rule be included in ESLint (instead of a plugin)?**

I cannot think of a valid use case why someone would want to define a getter that did not return a value and it can be easily overlooked when reviewing a long list of getter/setters. I think the problem is generic enough that it should be included by ESLint.

### PR DESCRIPTION
**Pull request by Aladdin-ADD**
fixes #8449 

**What is the purpose of this pull request? (put an "X" next to item)**
[ x] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-proposal.md))

**What changes did you make? (Give an overview)**
New rule: enforces that a return statement is present in property getters.

**Example code:**

```js
/*eslint getter-return: "error"*/

// Incorrect
class MyClass {
  get a() {
     this.retreiveVal('a');
  }
}

function MyConstructorFunction() {
  Object.defineProperty(this, 'a', {
    get: function () {
      this.retreiveVal('a');
    }
  });
}


// Correct
class MyClass {
  get a() {
     return this.retreiveVal('a');
  }
}

function MyConstructorFunction() {
  Object.defineProperty(this, 'a', {
    get: function () {
       return this.retreiveVal('a');
    }
  });
}
```

**Is there anything you'd like reviewers to focus on?**

+ I have confirmed the test coverage is 100%:
> 100% Statements 19/19 100% Branches 16/16 100% Functions 7/7 100% Lines 19/19
 
so the question is whether these test cases are fit for?

+ for code like this:
```js
  var foo = { get: function () {bar();} };
```
it is not a `getter` function, but naming a method `get` that does not return a value, is so confusing, causing an error may be somewhat reasonable.

+  My English is not so good(as you can see ^-^). More attention can be attached to the docs and comments.

### CODE REVIEW

**Comment by ilyavolodin** [comment]
@Aladdin-ADD Could you update the first post with the template information please? For example, is there an issue where this change was discussed?

**Comment by Aladdin-ADD** [comment]
sorry for that, link to the issue #8449 , as it is WIP, so...
I will add more info later.

File: tests/lib/rules/getter-return.js [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+        { code: "Object.defineProperty(foo, \"bar\", { get: function () {return true}})" },
+
+        { code: "var get = function(){};" }
+    ],
+
+    invalid: [
+
+        // TODO: why data: { name: "getter 'bar'"} is not working?
+        // test obj: get
+        { code: "var foo = { get bar() {return;} }", parserOptions, errors: [{ message: noReturnMessage, data: { name: "getter 'bar'" } }] },
```
Comment (Aladdin-ADD) [inline tests/lib/rules/getter-return.js]: SOS..can someone explain why this data cannot work?? @soda0289 @platinumazure

File: tests/lib/rules/getter-return.js [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+        { code: "var foo = { get bar() {return; ;} };", parserOptions, errors: [{ message: noReturnMessage, data: { name: "getter 'bar'" } }] },
+        { code: "var foo = { get bar() {return 1; return;} };", parserOptions, errors: [{ message: noReturnMessage, data: { name: "getter 'bar'" } }] },
+        { code: "var foo = { get bar() {return; return 1;} };", parserOptions, errors: [{ message: noReturnMessage, data: { name: "getter 'bar'" } }] },
+        { code: "var foo = { get bar(){if(bar) {return true;}} };", parserOptions, errors: [{ message: noLastReturnMessage, data: { name: "getter 'bar'" } }] },
+        { code: "var foo = { get bar(){if(bar) {return true;} return;} };", parserOptions, errors: [{ message: noReturnMessage, data: { name: "getter 'bar'" } }] },
+        { code: "var foo = { get bar(){if(bar) {return true;} ;} };", parserOptions, errors: [{ message: noLastReturnMessage, data: { name: "getter 'bar'" } }] },
+        { code: "var foo = { get bar(){if(bar) {return;} return true;} };", parserOptions, errors: [{ message: noReturnMessage, data: { name: "getter 'bar'" } }] },
```
Comment (ljharb) [inline tests/lib/rules/getter-return.js]: Why is this invalid? `undefined` is a valid return value for a getter.

File: tests/lib/rules/getter-return.js [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+
+        // test class: get
+        { code: "class foo { get bar(){} }", parserOptions, errors: [{ message: noReturnMessage, data: { name: "getter 'bar'" } }] },
+        { code: "class foo { get bar(){return;} }", parserOptions, errors: [{ message: noReturnMessage, data: { name: "getter 'bar'" } }] },
+        { code: "class foo { get bar(){if(bar) {return true;}} }", parserOptions, errors: [{ message: noLastReturnMessage, data: { name: "getter 'bar'" } }] },
+        { code: "class foo { get bar(){if(bar) {return;} return true;} }", parserOptions, errors: [{ message: noReturnMessage, data: { name: "getter 'bar'" } }] },
```
Comment (ljharb) [inline tests/lib/rules/getter-return.js]: Why is this invalid? `undefined` is a valid return value for a getter.

File: tests/lib/rules/getter-return.js [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+
+        // add object.defineProperty
+        { code: "Object.defineProperty(foo, \"bar\", { get: function (){return;}});", errors: [{ message: "Expected to return a value in method 'get'.", data: { name: "getter 'bar'" } }] },
+        { code: "Object.defineProperty(foo, \"bar\", { get: function (){if(bar) {return true;}}});", errors: [{ message: "Expected to return a value at the end of method 'get'.", data: { name: "getter 'bar'" } }] },
+        { code: "Object.defineProperty(foo, \"bar\", { get: function (){if(bar) {return;} return true;}});", errors: [{ message: "Expected to return a value in method 'get'.", data: { name: "getter 'bar'" } }] }
```
Comment (ljharb) [inline tests/lib/rules/getter-return.js]: Why is this invalid? `undefined` is a valid return value for a getter.

**Comment by platinumazure** [comment]
Is the goal of this rule to detect undefined returns, or to detect a lack of ReturnStatement nodes in a code path?

My assumption/preference would be the latter, in line with @ljharb's questions on the PR. But if I missed some discussion/conversation around this, please let me know.

File: tests/lib/rules/getter-return.js (hunk 2 above)
Comment (Aladdin-ADD) [inline tests/lib/rules/getter-return.js]: it is working in the same way as the rule `array-callback-return`, if we only detect a lack of ReturnStatement nodes in a code path,
```js
get foo(){
    //...
    return;
}
```
will be valid. it is werid..

**Comment by Aladdin-ADD** [comment]
@platinumazure @ljharb if we only detect a lack of ReturnStatement nodes in a code path,
```js
get foo(){
    //...
    return;
}
```
will be valid. it is werid..
so this `return` means `return a value explicitly`.

**Comment by soda0289** [comment]
If the rule is to enforce an explicit return in getters, maybe it should be called: `no-implicit-getter-return`

**Comment by platinumazure** [comment]
Or maybe we could add an option allowing users to choose which behavior
they want?

File: tests/lib/rules/getter-return.js (hunk 2 above)
Comment (ljharb) [inline tests/lib/rules/getter-return.js]: Why should it? There's already a separate rule to handle that (consistent-return) if that's desired.

**Comment by Aladdin-ADD** [comment]
as it is very similar to `array-callback-return`, for consistency consideration, we better open a new issue to talk about adding the new option.

**Comment by platinumazure** [comment]
I don't see why we can't add an option in the new rule, but if we definitely can't have an option in this initial implementation, then my preference would be for the rule to be as tolerant as possible and consider `return;` as a valid return statement for the purpose of this rule.

**Comment by Aladdin-ADD** [comment]
I mean maybe we should also add the option to `array-callback-return`.
updated the comment.sorry for inaccurate representation~~

**Comment by not-an-aardvark** [comment]
My opinion is that we should be consistent with `array-callback-return` and raise an error for `return;`. Usually, when I see `return;` it's being used for control flow, not returning `undefined`.

**Comment by ljharb** [comment]
fwiw, I use `return` exclusively over `return undefined` because `undefined` isn't a keyword, and isn't safe to rely on in non-app code (and `return void 0` is ugly).

The stated purpose of the rule is "enforces that a return statement is present in property getters.". `return;` is a return statement. If I'm typing it, I mean it - the rule is meant to warn me when I *omit* an explicit return.

**Comment by kaicataldo** [comment]
I'm in agreement with @ljharb. To me it feels like a bit of an overreach to disallow `return;`.

**Comment by Aladdin-ADD** [comment]
I'm not sure which behavior should be default.
while writing `return;` a developer may forget to return a value. So if not too much case `return undefined` in getter. I suggest this not be default. you recon? @ljharb

**Comment by ljharb** [comment]
`return;` *is* returning a value.

**Comment by not-an-aardvark** [comment]
It's true that `return;` implicitly returns a value, but omitting a return statement also implicitly returns a value. If the rule reports the latter and not the former, it will conflict with [`no-useless-return`](http://eslint.org/docs/rules/no-useless-return).

**Comment by ljharb** [comment]
@not-an-aardvark i don't see why it would *conflict* - it's more that this rule would allow `return;`, which `no-useless-return` could choose to handle - much like `one-var` might allow one var, but `no-var` would warn on it.

More to the point, *this* rule *should not* be concerning itself with "useless" returns, since that's the job of `no-useless-return`. It should ensure there's an explicit `return` keyword in a getter, nothing more.

**Comment by not-an-aardvark** [comment]
They would conflict because using `return;` at the end of a getter function would trigger `no-useless-return`, whereas omitting `return;` would trigger this rule. This would result in a situation where the user can't satisfy both rules (aside from using `return undefined`, but that's a separate issue).

**Comment by ljharb** [comment]
Right, in the case where both rules exist, `return undefined;` is absolutely the proper solution. I'm not seeing a conflict.

**Comment by soda0289** [comment]
I think `return undefined;` would conflict with the `no-undefined` rule

**Comment by platinumazure** [comment]
I think we can agree that certain rules might conflict with each other.
Sometimes users can get themselves into a no-win situation by enabling the
wrong combination of rules with the wrong options. I think if two rules
conflict with each other under *some* combination of options, that can be
okay as long as there are option sets that *don't* conflict. So let's
please not get caught up in possible conflicts, but rather design a
solution around every conflict we identify.

Here's an example. `return undefined;` can indeed conflict with
`no-undefined` rule. This suggests to me that this new rule either needs an
option to allow `return;` or should encourage users to use `return void 0;`.

I'm sure similar arguments could be created for `no-useless-return` when
users use `return;`. This would again lead me to suggest an option to
tolerate or forbid `return;`.

Do we need to consider, as separate issues, the idea of enhancing other
rules to avoid conflicts with this new rule?

I think by thinking about all the rules holistically, we will arrive at a
sensible design with only user-created conflicts, rather than inherent
design conflicts.

**Comment by soda0289** [comment]
Do you think we should make the default configuration as conflict free as possible? Maybe not allowing `return;` and `return undefined;` unless some specific options are enabled in this rule.

**Comment by Aladdin-ADD** [comment]
from my personal view, conflicts should be avoid as far as possible. But with more and more rules added, it is somewhat acceptable.
 **the rule's reasonability is in the first place. that's we should focus on.**

**Comment by ljharb** [comment]
As quoted in the OP: "New rule: enforces that a return statement is present in property getters.". `return;` is a return statement.

It should be left to `no-undefined` and/or `no-useless-return` to handle forbidding or requiring `return;` - I agree with https://github.com/eslint/eslint/pull/8460#issuecomment-300054181 in that adding options to other rules is preferable to making this rule overreach.

**Comment by Aladdin-ADD** [comment]
Is there a consensus with the members? no rush, but...(eh, my English... 😅 )

File: docs/rules/getter-return.md [hunk 5]
```diff
@@ -0,0 +1,80 @@
+# Enforces that a return statement is present in property getters (getter-return)
+
+The get syntax binds an object property to a function that will be called when that property is looked up. It is firstly introduced in ECMAScript 5:
```
Comment (soda0289) [inline docs/rules/getter-return.md]: I think the last sentence could read better. Maybe something like:
It was first introduced in ECMAScript 5

**Comment by ljharb** [comment]
I think that the current behavior is fine but that the option should be disabled by default - separately, "absence" and "false" should be equivalent, so if it's going to warn on `return;` by default, the option should be called "allowImplicit" or similar - boolean options should always be named such that they default to false.

File: lib/rules/getter-return.js [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+            description: "enforce `return` statements in getters",
+            category: "Possible Errors",
+            recommended: false
+        },
+        fixable: null,
+        schema: [
+            {
+                type: "object",
+                properties: {
+                    allowImplicit: {
```
Comment (ljharb) [inline lib/rules/getter-return.js]: Please add documentation for this option :-)

File: lib/rules/getter-return.js [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+    meta: {
+        docs: {
+            description: "enforce `return` statements in getters",
+            category: "Possible Errors",
+            recommended: false
+        },
+        fixable: null,
+        schema: [
+            {
+                type: "object",
```
Comment (ljharb) [inline lib/rules/getter-return.js]: this object should probably have `additionalProperties: false` set on it

File: lib/rules/getter-return.js [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+                funcInfo = {
+                    upper: funcInfo,
+                    codePath,
+                    hasReturn: false,
+                    shouldCheck:
+                        TARGET_NODE_TYPE.test(node.type) &&
+                        node.body.type === "BlockStatement" &&
+
+                        // check if it is a "getter", or a method named "get".
+                        (parent.kind === "get" || astUtils.getStaticPropertyName(parent) === "get"),
```
Comment (Aladdin-ADD) [inline lib/rules/getter-return.js]: for code like this:
```js
foo.__defineGetter__("bar", function baz() { return ""; });
```
`parent.kind` is not "get", is this a bug? @ljharb

Comment (ljharb) [inline lib/rules/getter-return.js]: It's a method named `__defineGetter__`; i wouldn't expect the parser to treat it as a syntactic getter.

File: tests/lib/rules/getter-return.js [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+    valid: [
+
+        // test obj: get, option: {allowImplicit: false}
+        { code: "var foo = { get bar(){return true;} };" },
+        { code: "var foo = { get bar(){return;} };" },
+        { code: "var foo = { bar: function(){return true;} };" },
+        { code: "var foo = { bar: function(){return;} };" },
+        { code: "var foo = { bar(){return true;} };", parserOptions },
+        { code: "var foo = { bar(){return;} };", parserOptions },
+        { code: "var foo = { bar(){~function (){}();return;} };", parserOptions },
```
Comment (kaicataldo) [inline tests/lib/rules/getter-return.js]: I actually have never seen this before. Mind explaining it to me? :)

Comment (Aladdin-ADD) [inline tests/lib/rules/getter-return.js]: `bar` is a normal method, not a getter. so a valid case.

Comment (ljharb) [inline tests/lib/rules/getter-return.js]: `~function(){}` is one of many tricks to make it immediately invokeable; you can also use `!`, or `delete`, or `void`, etc - but the most common (and only sensible) approach is wrapping the function in parens.

Comment (Aladdin-ADD) [inline tests/lib/rules/getter-return.js]: based on above case, I just add a IIFE, to test a nested function.

Comment (kaicataldo) [inline tests/lib/rules/getter-return.js]: Thank you for the explanation! This is why I love working on ESLint - I get to learn all sorts of things I wouldn't encounter at my day job :)

Comment (Aladdin-ADD) [inline tests/lib/rules/getter-return.js]: I find a mistake, the default option is allowImplicit : false, so it means we cannot return a value implicitly. in this case, should cause an error. thoughts? @kaicataldo @ljharb

Comment (ljharb) [inline tests/lib/rules/getter-return.js]: If that's the case then a number of these test cases should be invalid.

Comment (kaicataldo) [inline tests/lib/rules/getter-return.js]: Yeah, I'm a bit confused - line 36, for instance, should be a failing test, right?

File: lib/rules/getter-return.js [hunk 10]
```diff
@@ -0,0 +1,153 @@
...
+                funcInfo = funcInfo.upper;
+            },
+
+            // Checks the return statement is valid.
+            ReturnStatement(node) {
+                if (funcInfo.shouldCheck) {
+                    funcInfo.hasReturn = true;
+
+                    // if allowImplicit: true, should also check node.argument
+                    if (options.allowImplicit && !node.argument) {
```
Comment (kaicataldo) [inline lib/rules/getter-return.js]: I think this actually should be 
```js
// Checks the return statement is valid.
ReturnStatement(node) {
    if (funcInfo.shouldCheck) {
        funcInfo.hasReturn = true;

        // if allowImplicit: true, should also check node.argument
        if (!options.allowImplicit && !node.argument) {
            context.report({
                node,
                message: "Expected to return a value in {{name}}.",
                data: {
                    name: astUtils.getFunctionNameWithKind(funcInfo.node)
                }
            });
        }
    }
},
```

Comment (kaicataldo) [inline lib/rules/getter-return.js]: To see this behavior, try the following test with the current behavior:
```js
// test option: {allowImplicit: true}
{ code: "var foo = { get bar(){return;} };", options },
```
I think we should add this as a regression test once this is fixed. Good catch on this!

Comment (Aladdin-ADD) [inline lib/rules/getter-return.js]: yes, in previous implemnet, the default option is `allowImplicit`, seems I forgot to change this.
don't worry. I will fix it asap.

Comment (kaicataldo) [inline lib/rules/getter-return.js]: Thanks for working on this :)

Comment (Aladdin-ADD) [inline lib/rules/getter-return.js]: since we use --squash, so there is no need to rebase?

Comment (kaicataldo) [inline lib/rules/getter-return.js]: Yes, no worries about having multiple commits in a PR

File: tests/lib/rules/getter-return.js [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+        // test option: {allowImplicit: true}
+        { code: "var foo = { get bar() {return;} };", options },
+        { code: "var foo = { get bar(){return true;} };", options },
+        { code: "class foo { get bar(){return true;} }", options, parserOptions },
+        { code: "class foo { get bar(){return;} }", options, parserOptions },
+        { code: "Object.defineProperty(foo, \"bar\", { get: function () {return true;}});", options },
+        { code: "Object.defineProperies(foo, { bar: { get: function () {return true;}} });", options },
+        { code: "var foo = { get bar() {return;} };", options, parserOptions },
+        { code: "var foo = { get bar(){if(bar) {return;} return true;} };", options, parserOptions },
+        { code: "class foo { get bar(){return;} }", options, parserOptions },
```
Comment (kaicataldo) [inline tests/lib/rules/getter-return.js]: Duplicate of line 53

Comment (ljharb) [inline tests/lib/rules/getter-return.js]: lol eslint core should have a custom lint rule to catch duplicate rule test cases

Comment (kaicataldo) [inline tests/lib/rules/getter-return.js]: That, my friend, is a fantastic idea

File: tests/lib/rules/getter-return.js [hunk 12]
```diff
@@ -0,0 +1,91 @@
...
+        { code: "Object.defineProperies(foo, { bar: { get: function () { ~function (){ return true; }(); return true;}} });" },
+
+        // test option: {allowImplicit: true}
+        { code: "var foo = { get bar() {return;} };", options },
+        { code: "var foo = { get bar(){return true;} };", options },
+        { code: "class foo { get bar(){return true;} }", options, parserOptions },
+        { code: "class foo { get bar(){return;} }", options, parserOptions },
+        { code: "Object.defineProperty(foo, \"bar\", { get: function () {return true;}});", options },
+        { code: "Object.defineProperies(foo, { bar: { get: function () {return true;}} });", options },
+        { code: "var foo = { get bar() {return;} };", options, parserOptions },
```
Comment (kaicataldo) [inline tests/lib/rules/getter-return.js]: I'm confused by this test - it seems to be the same as the one on line 50 but has the additional `parserOptions`, which seems irrelevant. Am I missing something?

File: tests/lib/rules/getter-return.js (hunk 11 above)
Comment (not-an-aardvark) [inline tests/lib/rules/getter-return.js]: https://github.com/not-an-aardvark/eslint-plugin-eslint-plugin/issues/5

**Comment by Aladdin-ADD** [comment]
friendly ping. @kaicataldo

File: docs/rules/getter-return.md [hunk 13]
```diff
@@ -0,0 +1,86 @@
...
+        return "nicholas";
+    }
+}
+```
+
+## Options
+
+This rule has an object option:
+
+* `"allowImplicit": false` (default) disallows implicitly return a value.
```
Comment (not-an-aardvark) [inline docs/rules/getter-return.md]: Can you please add examples for this option?

Also, I think it would be a good idea to clarify the option description, e.g. "...disallows implicitly returning `undefined` with a `return;` statement".

File: docs/rules/getter-return.md [hunk 14]
```diff
@@ -0,0 +1,86 @@
...
+
+## When Not To Use It
+
+If your project will not be using getter you do not need this rule.
```
Comment (not-an-aardvark) [inline docs/rules/getter-return.md]: Nitpick: This should say "if your project will not be using *getters*".

File: lib/rules/getter-return.js [hunk 15]
```diff
@@ -0,0 +1,153 @@
...
+            // Stacks this function's information.
+            onCodePathStart(codePath, node) {
+                const parent = node.parent;
+
+                funcInfo = {
+                    upper: funcInfo,
+                    codePath,
+                    hasReturn: false,
+                    shouldCheck:
+                        TARGET_NODE_TYPE.test(node.type) &&
```
Comment (not-an-aardvark) [inline lib/rules/getter-return.js]: Nitpick: Since `TARGET_NODE_TYPE` only has one type of function, can you change this to just be `node.type === "FunctionExpression"`?

File: lib/rules/getter-return.js [hunk 16]
```diff
@@ -0,0 +1,153 @@
...
+
+/**
+ * Gets a readable location.
+ *
+ * - FunctionExpression -> the function name or `function` keyword.
+ *
+ * @param {ASTNode} node - A function node to get.
+ * @returns {ASTNode|Token} The node or the token of a location.
+ */
+function getLocation(node) {
```
Comment (not-an-aardvark) [inline lib/rules/getter-return.js]: Can you rename this function to something like `getId`? I think the name `getLocation` is misleading because it's not actually returning a location.

File: tests/lib/rules/getter-return.js [hunk 17]
```diff
@@ -0,0 +1,97 @@
...
+// Tests
+//------------------------------------------------------------------------------
+
+const ruleTester = new RuleTester();
+
+// data is not working, so specify a name: "getter 'bar'"
+const name = "getter 'bar'";
+const noReturnMessage = `Expected to return a value in ${name}.`;
+const noLastReturnMessage = `Expected to return a value at the end of ${name}.`;
+const parserOptions = { ecmaVersion: 6 };
```
Comment (not-an-aardvark) [inline tests/lib/rules/getter-return.js]: Nitpick: Instead of using `parserOptions` for each test, you can set defaults:

```js
const ruleTester = new RuleTester({ parserOptions: { ecmaVersion: 6 } });
```

Then you won't need to repeat `parserOptions` in so many places.

File: docs/rules/getter-return.md (hunk 14 above)
Comment (ljharb) [inline docs/rules/getter-return.md]: I might even call them "ES5 property getters" to distinguish them from generic getFoo methods.

File: lib/rules/getter-return.js [hunk 18]
```diff
@@ -0,0 +1,151 @@
...
+         * @param {ASTNode} node - A node to check.
+         * @returns {void}
+         */
+        function checkLastSegment(node) {
+            if (funcInfo.shouldCheck &&
+                funcInfo.codePath.currentSegments.some(isReachable)
+            ) {
+                context.report({
+                    node,
+                    loc: getId(node, context.getSourceCode()).loc.start,
```
Comment (platinumazure) [inline lib/rules/getter-return.js]: I'm a bit confused by this call, which passes `context.getSourceCode()` as an argument but which does not match any formal parameter defined on the function. I think the second argument can probably be removed?

File: lib/rules/getter-return.js [hunk 19]
```diff
@@ -0,0 +1,151 @@
...
+                    message: funcInfo.hasReturn
+                        ? "Expected to return a value at the end of {{name}}."
```
Comment (platinumazure) [inline lib/rules/getter-return.js]: Not sure this message fits for some cases:

```js
{
    get foo() {
        if (bar) {
            return true;
        } else if (baz) {
            doSomething();
            // missing return here
        } else {
            return false;
        }
    }
}
```

In the example above, the return isn't missing at the end of the function.

I think a better message might be something like this: "Expected to return a value from all code paths of {{name}}." But there might be even better options.

File: tests/lib/rules/getter-return.js [hunk 20]
```diff
@@ -0,0 +1,96 @@
...
+        { code: "Object.defineProperty(foo, \"bar\", { get: function () {return true;}});", options },
+        { code: "Object.defineProperies(foo, { bar: { get: function () {return true;}} });", options },
+        { code: "Object.defineProperty(foo, \"bar\", { get: function (){return;}});", options },
+        { code: "Object.defineProperies(foo, { bar: { get: function () {return;}} });", options },
+
+        // not getter.
+        { code: "var get = function(){};" },
+        { code: "var get = function(){ return true; };" },
+        { code: "var foo = { bar: function(){} };" },
+        { code: "var foo = { bar: function(){ return true; } };" },
```
Comment (platinumazure) [inline tests/lib/rules/getter-return.js]: This test seems identical to line 33?

File: tests/lib/rules/getter-return.js [hunk 21]
```diff
@@ -0,0 +1,96 @@
...
+        { code: "var foo = { bar(){} };" },
+        { code: "var foo = { bar(){ return true; } };" }
+    ],
+
+    invalid: [
+
+        // test obj: get, option: {allowImplicit: false}
+        { code: "var foo = { get bar() {} };", errors: [{ message: noReturnMessage }] },
+        { code: "var foo = { get bar(){if(bar) {return true;}} };", errors: [{ message: noLastReturnMessage }] },
```
Comment (platinumazure) [inline tests/lib/rules/getter-return.js]: Wondering if it would be less confusing to use `if(baz)`, like in later tests, to avoid confusion around the variable name being the same as the property.

File: lib/rules/getter-return.js (hunk 19 above)
Comment (not-an-aardvark) [inline lib/rules/getter-return.js]: How about "Expected {{name}} to always return a value."?

**Comment by Aladdin-ADD** [comment]
updated. and reorganized the tests, to avoid duplicate.  :(

**Comment by kaicataldo** [comment]
Thanks for contributing to ESLint!
